contrib/functions/go/annotate-apply-time-mutations/pkg/comment_scanner.go

```diff
@@ -0,0 +1,180 @@
+package pkg
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/cli-utils/pkg/object/mutation"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	tokenPattern              = "${ref%d}"
+	sourceRegex               = `\${(?P<group>[^/]+)/((?P<version>[^/]+)/)?namespaces/(?P<namespace>[^/]+)/(?P<kind>[^/]+)/(?P<name>[^:]+):(?P<path>[^}]+)}`
+	onlySourceRegex           = "^" + sourceRegex + "$"
+	mutationCommentIdentifier = "# apply-time-mutation: "
+	ResourceGroup             = "function.kpt.dev"
+	ResourceVersion           = "v1alpha1"
+)
+
+var (
+	resourceReferencePattern = regexp.MustCompile(sourceRegex)
+	onlyReferencePattern     = regexp.MustCompile(onlySourceRegex)
+)
+
+type CommentScanner struct {
+	// ObjMeta is a reference to the object using metadata
+	ObjMeta yaml.ResourceIdentifier
+	// ObjFile is a reference to the object using file and object index
+	ObjFile framework.File
+}
+
+type ScanResult struct {
+	Path         string
+	Value        interface{}
+	Comment      string
+	Substitution mutation.FieldSubstitution
+}
+
+// Scan searches for mutation markup comments and parses them as substitutions.
+func (cs *CommentScanner) Scan(obj *yaml.RNode, results map[string]ScanResult, fieldPath ...interface{}) error {
```

Comment:
It's used to build a JSONPath expression recursively. The Substitution spec requires JSONPath.

`framework.Field.Path` in `framework.Result.Field` also requires a string, tho it might not actually be JSONPath. It seems to be a concatenation of map keys delimited by periods without any list indexes... That format seems pretty useless, but it's not clear if it's used for anything other than printing to the user.

Reply:
Digging deeper, the Field.Path in the framework is exclusively used to generate stringified user output, not for looking up the field in any way. So the format of the Path is probably irrelevant, except for consistency. And since the convention sucks (always  period delimited, even if the field has period or quote or bracket in it... and skips all list indexes...) I think it's probably fine to just set it to the JSONPath value I need for the apply-time-mutation annotation anyway.

But because you asked nicely, I wrote a simple PathBuilder to pass around and generate the JSONPath on-demand instead of the vararg of interface{}. It also incidentally simplifies the errors and logging with path in them.